File: embedding.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAX_TOKENS 8191

typedef struct {
    double *embedding;
    int index;
} OpenAIEmbedding;

typedef struct {
    int count;
    int max_tokens;
    char model_name[256];
    OpenAIEmbedding embedding;
} EmbeddingContainer;

double generate_random() {
    return ((double)rand() / RAND_MAX) * 2.0 - 1.0;
}

void generate_embedding_vector(double *embedding, int num_dimensions) {
    for (int i = 0; i < num_dimensions; i++) {
        embedding[i] = generate_random();
    }
}

int generate_tokens(int max_tokens) {
    return rand() % max_tokens + 1;
}

void init_embedding_container(EmbeddingContainer *container, int num_dimensions, int max_tokens, const char *model_name, int index) {
    container->count = 1;
    container->max_tokens = max_tokens;
    strncpy(container->model_name, model_name, sizeof(container->model_name));
    generate_embedding_vector(container->embedding.embedding, num_dimensions);
    container->embedding.index = index;
}

OpenAIEmbedding *to_dict(EmbeddingContainer *container) {
    return &container->embedding;
}

void save_to_file(EmbeddingContainer *container, const char *file_path) {
    FILE *file = fopen(file_path, "wb");
    if (file == NULL) {
        return;
    }
    fprintf(file, "{\"data\":[{\"embedding\":[");
    for (int i = 0; i < container->count; i++) {
        fprintf(file, "%f,", container->embedding.embedding[i]);
    }
    fprintf(file, "],\"index\":%d,\"object\":\"embedding\"}],\"model\":\"%s\",\"object\":\"list\",\"usage\":{\"prompt_tokens\":%d,\"total_tokens\":%d}}", container->embedding.index, container->model_name, container->max_tokens, container->max_tokens);
    fclose(file);
}

File: embedding.py
import random
import orjson

from memory_profiler import profile
from typing import List, Dict, Tuple, Union

OpenAIEmbedding = Dict[
    str, Union[str, int, List[Dict[str, Union[int, str, List[float]]]]]
]


class EmbeddingContainer:
    __slots__: Tuple[str, ...] = (
        "_dims",
        "_model_name",
        "_i",
        "_max_tokens",
        "_vec",
        "_tokens",
    )

    def __init__(
        self,
        num_dimensions: int,
        max_tokens: int,
        model_name: str = "text-embedding-ada-002",
        index: int = 0,
    ):
        self._dims: int = num_dimensions
        self._max_tokens: int = max_tokens
        self._model_name: str = model_name
        self._i: int = index
        self._tokens: int = EmbeddingContainer.generate_tokens(self._max_tokens)
        self._vec: List[float] = EmbeddingContainer.generate_embedding_vector(
            self._dims
        )

    @staticmethod
    def generate_embedding_vector(num_dimensions: int) -> List[float]:
        return [random.uniform(-1, 1) for _ in range(num_dimensions)]

    @staticmethod
    def generate_tokens(max_tokens: int) -> int:
        return random.randint(1, max_tokens)

    def to_dict(self) -> OpenAIEmbedding:
        return {
            "data": [{"embedding": self._vec, "index": self._i, "object": "embedding"}],
            "model": self._model_name,
            "object": "list",
            "usage": {"prompt_tokens": self._tokens, "total_tokens": self._tokens},
        }

    def save_to_file(self, file_path: str) -> None:
        with open(file_path, "wb") as f:
            f.write(orjson.dumps(self.to_dict()))
